metricbeat/module/kubernetes/events/eventwatcher.go

```diff
@@ -0,0 +1,116 @@
+package events
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/ericchiang/k8s"
+	corev1 "github.com/ericchiang/k8s/api/v1"
+)
+
+// EventWatcher is a controller that synchronizes Pods.
+type EventWatcher struct {
+	kubeClient          *k8s.Client
+	namespace           string
+	syncPeriod          time.Duration
+	eventQueue          chan *Event
+	lastResourceVersion string
+	ctx                 context.Context
+	stop                context.CancelFunc
+}
+
+// NewEventWatcher initializes the watcher client to provide a local state of
+// pods from the cluster (filtered to the given host)
+func NewEventWatcher(kubeClient *k8s.Client, syncPeriod time.Duration, namespace string) *EventWatcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &EventWatcher{
+		kubeClient:          kubeClient,
+		namespace:           namespace,
+		syncPeriod:          syncPeriod,
+		eventQueue:          make(chan *Event, 10),
+		lastResourceVersion: "0",
+		ctx:                 ctx,
+		stop:                cancel,
+	}
+}
+
+// watchEvents watches on the Kubernetes API server and puts them onto a channel.
+// watchEvents only starts from the most recent event.
+func (p *EventWatcher) watchEvents() {
+	for {
+		//To avoid writing old events, list events to get last resource version
+		events, err := p.kubeClient.CoreV1().ListEvents(
+			p.ctx,
+			p.namespace,
+		)
+
+		if err != nil {
+			//if listing fails try again after sometime
+			logp.Err("kubernetes: List API error %v", err)
+			time.Sleep(time.Second)
```

Comment:
Based on what did you select this timeout? Perhaps a comment could be helpful here.